tar/lib/src/reader.dart

```diff
@@ -0,0 +1,465 @@
+// Copyright 2020
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'dart:async';
+
+import 'package:collection/collection.dart';
+import 'package:meta/meta.dart';
+
+import 'constants.dart';
+import 'exceptions.dart';
+import 'format.dart';
+import 'header.dart';
+import 'header_impl.dart';
+import 'sparse_entry.dart';
+import 'stream.dart';
+import 'utils.dart';
+
+/// [TarReader] provides sequential access to the TAR files in a TAR archive.
+/// It is designed to read from a stream and to spit out substreams for
+/// individual file contents in order to minimize the amount of memory needed
+/// to read each archive where possible.
+@sealed
+class TarReader {
+  /// A chunked stream iterator to enable us to get our data.
+  final ChunkedStreamIterator<int> _chunkedStream;
+
+  /// The [TarHeader] for the current file. Completes when a call to `next`
+  /// completes as well.
+  ///
+  /// Is `null` before the first call to `next` and after a call to `next`
+  /// completes with a `false` result.
+  Future<TarHeader> get header => _header;
+  Future<TarHeader> _header;
```

Comment:
Should we consider making `header == null` while `next()` has not returned... once `next()` returns then we unassign this for `null`.

Reply:
hm - with the current implementation, header is a `Future` that is set to a new `Completer`'s future when `next` is called, then completed when `next()` returns. Making it `null` instead might make it less friendly, since currently the `Future` reminds them to `await` for the result, while if we changed it to `null`, they might use the `null` in other areas which leads to errors that might be harder to debug?

Reply:
Well, you're supposed to be awaiting the `next()` call. So accessing `header` or `stream` before that completes seems unnecessary.

But maybe we should wait a bit, and see if Simon wants to take this somewhere.

Reply:
> But maybe we should wait a bit, and see if Simon wants to take this somewhere.

I do! I started to integrate the code for tar formats and header parsing from here (in https://github.com/simolus3/tar/commit/583c3398b46f6a974634f04697579e4e2bfc73af). I'll copy the remaining parser and tests soon.